### PR TITLE
Uf 3265 add parking permit renewal rule

### PR DESCRIPTION
**Pull request by josc04**
- Added rule for renewal of parking permit for driver
- Added rule for renewal of parking permit for passenger
- Refactored rule logic to encapsule handling in each rule member where each rule is independent of other rule members
- Added sanity checks for needed parameters for each rule
- Moved rules package to not be inside camunda integration package
- Moved some other classes for better package structure
- Removed unnecessary service classes

### CODE REVIEW

File: src/main/java/se/sundsvall/processengine/parkingpermit/businessrules/BusinessRule.java [hunk 1]
```diff
@@ -0,0 +1,10 @@
+package se.sundsvall.processengine.parkingpermit.businessrules;
+
+interface BusinessRule {
+	boolean shouldBeApplied(ErrandDecisionWrapper errandDecision);
+	ErrandDecisionWrapper apply(ErrandDecisionWrapper errandDecision);
+    
+    static boolean isDecisionPresent(ErrandDecisionWrapper errandDecision) {
```
Comment (foobar-1337) [inline src/main/java/se/sundsvall/processengine/parkingpermit/businessrules/BusinessRule.java:7]: Indentation

File: src/main/java/se/sundsvall/processengine/parkingpermit/businessrules/DriverNewParkingPermitRule.java [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+		
+		return true;
+	}
+	
+	private static Optional<DecisionDTO> createDecisionBasedOnWalkingAbility(boolean walkingAbility, String maxWalkingDistance) {
+		if (!walkingAbility) {
+			return createDecision(RECOMMENDED, APPROVAL, APPROVAL_NO_WALKING_ABILITY);
+		} else if (parseInt(maxWalkingDistance) < RULE_WALKING_DISTANCE_LIMIT_METERS) {
+			return createDecision(RECOMMENDED, APPROVAL, APPROVAL_WALKING_DISTANCE_LESS_THAN_500_METERS);
+		} else {
```
Comment (foobar-1337) [inline src/main/java/se/sundsvall/processengine/parkingpermit/businessrules/DriverNewParkingPermitRule.java:81]: Redundant else

File: src/main/java/se/sundsvall/processengine/parkingpermit/businessrules/DriverRenewalParkingPermitRule.java [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+		}		
+		return true;
+	}
+	
+	private static Optional<DecisionDTO> createDecisionBasedOnWalkingAbility(boolean changedCircumstances, boolean walkingAbility, String maxWalkingDistance) {
+		if (!walkingAbility) {
+			return createDecision(RECOMMENDED, APPROVAL, changedCircumstances ? APPROVAL_NO_WALKING_ABILITY : APPROVAL_CIRCUMSTANCES_UNCHANGED);
+		} else if (parseInt(maxWalkingDistance) < RULE_WALKING_DISTANCE_LIMIT_METERS) {
+			return createDecision(RECOMMENDED, APPROVAL, changedCircumstances ? APPROVAL_WALKING_DISTANCE_LESS_THAN_500_METERS : APPROVAL_CIRCUMSTANCES_UNCHANGED);
+		} else {
```
Comment (foobar-1337) [inline src/main/java/se/sundsvall/processengine/parkingpermit/businessrules/DriverRenewalParkingPermitRule.java:81]: Redundant else

File: src/main/java/se/sundsvall/processengine/parkingpermit/businessrules/ErrandDecisionWrapper.java [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+	}
+
+	@Override
+	public int hashCode() {
+		return Objects.hash(decision, errand);
+	}
+
+	@Override
+	public boolean equals(Object obj) {
+		if (this == obj)
```
Comment (foobar-1337) [inline src/main/java/se/sundsvall/processengine/parkingpermit/businessrules/ErrandDecisionWrapper.java:55]: Use curly braces everywhere for increased readability?

File: src/main/java/se/sundsvall/processengine/parkingpermit/businessrules/PassengerNewParkingPermitRule.java [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+	}
+
+	private static Optional<DecisionDTO> createDecisionBasedOnDisabilityDuration(String disabilityDuration) {
+		if (Period.parse(disabilityDuration).getMonths() > RULE_DISABILITY_DURATION_LIMIT_MONTHS) {
+			return createDecision(RECOMMENDED, APPROVAL, APPROVAL_CANT_BE_ALONE_WHILE_PARKING);
+		}
+		return createDecision(RECOMMENDED, REJECTION, REJECTION_DISABILITY_ENDS_IN_6_MONTHS_OR_LESS);
+	}
+	
+	private static boolean hasWalkingAbility(String canBeAloneWhileParking, String maxWalkingDistance) {
```
Comment (foobar-1337) [inline src/main/java/se/sundsvall/processengine/parkingpermit/businessrules/PassengerNewParkingPermitRule.java:79]: I can't see the connection with walking-ability and the possibility to be left alone while parking?

Maybe a stupid question: but how can the "can-be-left-alone"-criteria affect the walking ability?

Refactor in order to make it more logical?

Comment (josc04) [reply src/main/java/se/sundsvall/processengine/parkingpermit/businessrules/PassengerNewParkingPermitRule.java:79]: The defined business rules (https://sundsvall.atlassian.net/wiki/spaces/SK/pages/1099595803/Aff+rsreglers+ans+kan+P-tillst+nd) states that if the person can't be left alone, the person must also be unable to walk a distance longer than 99 meters. 

Don't ask me why these are the rules, I just implement the rules accordingly to the demands. 🙂

Comment (foobar-1337) [reply src/main/java/se/sundsvall/processengine/parkingpermit/businessrules/PassengerNewParkingPermitRule.java:79]: That was not what I commented about. 
The business rule seems fair, but the method name "hasWalkingAbility" seems illogical (in mind mind).

I.e. 
"can-be-left-alone" affects walking-ability? Nope.
walking-ability affects "can-be-left-alone"? Yes.

If you agree to this, then you would probably also agree to refactor the method, in order to give it a more logical name?

Comment (josc04) [reply src/main/java/se/sundsvall/processengine/parkingpermit/businessrules/PassengerNewParkingPermitRule.java:79]: Sure, I misinterpreted your comment. Will try to refactor/rename method more describe what it's doing.

**Review by MattiasGradinApendo — APPROVED** [review]

**Comment by josc04** [comment]
Acted on all PR comments

**Review by foobar-1337 — APPROVED** [review]

**Pull request merged**